File: app/src/main/java/com/example/myapplication/services/MyService.java
package com.example.myapplication.services;

import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.media.MediaMetadata;
import android.media.session.MediaController;
import android.media.session.MediaSessionManager;
import android.media.session.PlaybackState;
import android.net.ConnectivityManager;
import android.os.Binder;
import android.os.IBinder;
import android.service.notification.NotificationListenerService;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.NotificationManagerCompat;

import com.example.myapplication.R;
import com.example.myapplication.activities.PlayedActivity;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.WeakHashMap;

public class MyService extends NotificationListenerService
        implements MediaSessionManager.OnActiveSessionsChangedListener,
        SharedPreferences.OnSharedPreferenceChangeListener {

    private static List<MediaController> mediaControllers = new ArrayList<>();
    private Map<MediaController, MediaController.Callback> controllerCallbacks = new WeakHashMap<>();
    MyBinder binder = new MyBinder();

    @Override
    public IBinder onBind(Intent intent) {

        return binder;
    }

    @Override
    public boolean onUnbind(Intent intent) {
        return super.onUnbind(intent);
    }

    @Override
    public void onCreate() {
        super.onCreate();

        ConnectivityManager connectivityManager =
                (ConnectivityManager) this.getSystemService(Context.CONNECTIVITY_SERVICE);

        MediaSessionManager mediaSessionManager =
                (MediaSessionManager)
                        getApplicationContext().getSystemService(Context.MEDIA_SESSION_SERVICE);

        ComponentName componentName = new ComponentName(this, this.getClass());
        List<MediaController> initialSessions = null;
        if (mediaSessionManager != null) {
            mediaSessionManager.addOnActiveSessionsChangedListener(this, componentName);
            initialSessions = mediaSessionManager.getActiveSessions(componentName);
        }
        onActiveSessionsChanged(initialSessions);

    }

    @Override
    public void onActiveSessionsChanged(@Nullable List<MediaController> activeMediaControllers) {

        for (MediaController controller : mediaControllers) {
            if(controllerCallbacks.size()>0){
                controller.unregisterCallback(controllerCallbacks.get(controller));
                controllerCallbacks.remove(controller);
            }
        }
        controllerCallbacks.clear();

        if (activeMediaControllers != null) {
            for (final MediaController controller : activeMediaControllers) {
                //String packageName = controller.getPackageName();

                MediaController.Callback callback =
                        new MediaController.Callback() {
                            @Override
                            public void onPlaybackStateChanged(@NonNull PlaybackState state) {
                                controllerPlaybackStateChanged(controller, state);
                            }

                            @Override
                            public void onMetadataChanged(MediaMetadata metadata) {
                                controllerMetadataChanged(controller, metadata);
                            }
                        };

                controllerCallbacks.put(controller, callback);
                controller.registerCallback(callback);

                // Media may already be playing - update with current state.

                controllerPlaybackStateChanged(controller, controller.getPlaybackState());
                controllerMetadataChanged(controller, controller.getMetadata());
            }
        }

        mediaControllers = activeMediaControllers;
    }

    public static boolean isNotificationAccessEnabled(Context context) {
        return NotificationManagerCompat.getEnabledListenerPackages(context)
                .contains(context.getPackageName());
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String s) {

    }

    private void controllerPlaybackStateChanged(MediaController controller, PlaybackState state) {
        if (state!=null && state.getState() != PlaybackState.STATE_PLAYING){
            NotificationManager notificationManager =
                    (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
            if (notificationManager != null) {
                notificationManager.cancel(0);
            }
        }
    }

    private static PlayedActivity act;

    private void controllerMetadataChanged(MediaController controller, MediaMetadata metadata) {

        notify(controller);
        if(metadata!=null && act!=null){
            act.doIt(metadata);
        }
    }

    public void notify(MediaController controller){
        if((controller.getMetadata()!=null)&&(controller.getPlaybackState()!=null))
            if(controller.getPlaybackState().getState() == PlaybackState.STATE_PLAYING) {
                final String str = controller.getMetadata().getString(MediaMetadata.METADATA_KEY_TITLE);

                Intent intent = new Intent(this, PlayedActivity.class);
                PendingIntent pendingIntent =
                        PendingIntent.getActivity(this, 0, intent, PendingIntent.FLAG_CANCEL_CURRENT);
                Notification.Builder notification =
                        new Notification.Builder(this)
                                .setSmallIcon(R.drawable.ic_menu_camera)
                                .setContentTitle("Now playing")
                                .setContentText(str)
                                .setOngoing(true)
                                .setCategory(Notification.CATEGORY_STATUS)
                                .setColor(Color.argb(125, 242, 72, 63))
                                .setContentIntent(pendingIntent);
                NotificationManager notificationManager = (NotificationManager) this.getSystemService(Context.NOTIFICATION_SERVICE);
                if (notificationManager != null) {
                    notificationManager.notify(0, notification.build());
                }
            }
    }

    public List<MediaController> getMediaControllers(){
        return mediaControllers;
    }

    public void setAct(PlayedActivity a){
        act = a;
    }

    public class MyBinder extends Binder{
       public MyService getService(){
            return MyService.this;
        }
    }
}

File: app/src/main/java/com/example/myapplication/services/MyService.kt
package com.example.myapplication.services

import android.app.NotificationManager
import android.app.PendingIntent
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.media.MediaMetadata
import android.media.session.MediaController
import android.media.session.MediaSessionManager
import android.media.session.PlaybackState
//import android.net.ConnectivityManager
import android.os.Binder
import android.os.IBinder
import android.service.notification.NotificationListenerService
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import com.example.myapplication.MyApp.Companion.channel_id1
import com.example.myapplication.MyApp.Companion.hist
import com.example.myapplication.R
import com.example.myapplication.activities.PlayedActivity
import java.util.*

class MyService: NotificationListenerService(), MediaSessionManager.OnActiveSessionsChangedListener {
    private val controllerCallbacks = WeakHashMap<MediaController, MediaController.Callback>()
    private var binder = MyBinder()

    override fun onBind(intent: Intent): IBinder? {

        return binder
    }

    override fun onCreate() {
        super.onCreate()

        //val connectivityManager = this.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        //connectivityManager.isDefaultNetworkActive
        val mediaSessionManager = applicationContext.getSystemService(Context.MEDIA_SESSION_SERVICE) as MediaSessionManager

        val componentName = ComponentName(this, this.javaClass)
        val initialSessions: List<MediaController>?

        mediaSessionManager.addOnActiveSessionsChangedListener(this, componentName)
        initialSessions = mediaSessionManager.getActiveSessions(componentName)

        onActiveSessionsChanged(initialSessions)

    }

    override fun onDestroy() {
        val notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        notificationManager.cancel(0)
        super.onDestroy()
    }

    override fun onActiveSessionsChanged(activeMediaControllers: List<MediaController>?) {

        for (controller in mediaControllers) {
            if (controllerCallbacks.size > 0) {
                controller.unregisterCallback(Objects.requireNonNull<MediaController.Callback>(controllerCallbacks[controller]))
                controllerCallbacks.remove(controller)
            }
        }
        controllerCallbacks.clear()
        val notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        notificationManager.cancel(0)

        if (activeMediaControllers != null) {
            for (controller in activeMediaControllers) {
                //val packageName = controller.packageName
                //if (packageName == "com.example.myapplication") continue
                val callback = object : MediaController.Callback() {
                    override fun onPlaybackStateChanged(state: PlaybackState?) {
                        controllerPlaybackStateChanged(controller, state)
                    }

                    override fun onMetadataChanged(metadata: MediaMetadata?) {
                        controllerMetadataChanged(controller, metadata)
                    }
                }

                controllerCallbacks[controller] = callback
                controller.registerCallback(callback)

                controllerPlaybackStateChanged(controller, controller.playbackState)
                //controllerMetadataChanged(controller, controller.metadata)
            }
        }

        mediaControllers = activeMediaControllers!!
    }

    private fun controllerPlaybackStateChanged(controller: MediaController, state: PlaybackState?) {
        controller.sessionActivity
        if (state != null && state.state != PlaybackState.STATE_PLAYING) {
            val notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            notificationManager.cancel(0)
        } else notify(controller)
    }

    private fun controllerMetadataChanged(controller: MediaController, metadata: MediaMetadata?) {

        notify(controller)
        if (metadata != null) {
            hist.add(name = metadata.getString(MediaMetadata.METADATA_KEY_TITLE),
                    album = metadata.getString(MediaMetadata.METADATA_KEY_ARTIST))
            act?.doIt(metadata)
        }
    }

    private fun notify(controller: MediaController) {
        if (controller.metadata != null && controller.playbackState != null)
            if (controller.playbackState!!.state == PlaybackState.STATE_PLAYING) {
                val str = controller.metadata!!.getString(MediaMetadata.METADATA_KEY_TITLE)

                val intent = Intent(this, PlayedActivity::class.java)
                val pendingIntent = PendingIntent.getActivity(this, 0, intent, PendingIntent.FLAG_CANCEL_CURRENT)
                val notification = NotificationCompat.Builder(this, channel_id1)
                        .setSmallIcon(R.drawable.ic_menu_camera)
                        .setContentTitle("Now playing")
                        .setContentText(str)
                        .setOngoing(true)
                        .setContentIntent(pendingIntent)
                val notificationManager = this.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
                notificationManager.notify(0, notification.build())
                //startForeground(0, notification.build())
            }
    }

    fun getMediaControllers(): List<MediaController> {
        return mediaControllers
    }

    fun setAct(a: PlayedActivity?) {
        act = a
    }

    inner class MyBinder: Binder() {
        val service: MyService
            get() = this@MyService
    }

    companion object {

        var mediaControllers: List<MediaController> = ArrayList()

        fun isNotificationAccessEnabled(context: Context): Boolean {
            return NotificationManagerCompat.getEnabledListenerPackages(context)
                    .contains(context.packageName)
        }

        private var act: PlayedActivity? = null
    }
}
